src/ext/legit.lisp

```diff
@@ -0,0 +1,507 @@
+;; Supercharge patch-mode with our keys.
+(define-major-mode legit-diff-mode lem-patch-mode:patch-mode
+    (:name "legit-diff"
+     :syntax-table lem-patch-mode::*patch-syntax-table*
+     :keymap *legit-diff-mode-keymap*)
+  (setf (variable-value 'enable-syntax-highlight) t))
+
+;; git commands.
+;; Some are defined on peek-legit too.
+(define-key *global-keymap* "C-x g" 'legit-status)
+(define-key *legit-diff-mode-keymap* "s" 'legit-stage-hunk)
+(define-key *legit-diff-mode-keymap* "u" 'legit-unstage-hunk)
+(define-key *legit-diff-mode-keymap* "n" 'legit-goto-next-hunk)
+(define-key *legit-diff-mode-keymap* "p" 'legit-goto-previous-hunk)
+
+(define-key *legit-diff-mode-keymap* "c" 'legit-commit)
+(define-key lem/peek-legit::*peek-legit-keymap* "c" 'legit-commit)
+
+(define-key lem/peek-legit::*peek-legit-keymap* "b b" 'legit-branch-checkout)
+(define-key *legit-diff-mode-keymap* "b b" 'legit-branch-checkout)
+(define-key lem/peek-legit::*peek-legit-keymap* "b c" 'legit-branch-create)
+(define-key *legit-diff-mode-keymap* "b c" 'legit-branch-create)
+;; push
+(define-key *legit-diff-mode-keymap* "P p" 'legit-push)
+(define-key lem/peek-legit::*peek-legit-keymap* "P p" 'legit-push)
+;; pull
+(define-key lem/peek-legit::*peek-legit-keymap* "F p" 'legit-pull)
+(define-key *legit-diff-mode-keymap* "F p" 'legit-pull)
+
+;; redraw everything:
+(define-key lem/peek-legit::*peek-legit-keymap* "g" 'legit-status)
+
+;; navigation
+(define-key *legit-diff-mode-keymap* "C-n" 'next-line)
+(define-key *legit-diff-mode-keymap* "C-p" 'previous-line)
+(define-key lem/peek-legit::*peek-legit-keymap* "M-n" 'legit-next-header)
+(define-key lem/peek-legit::*peek-legit-keymap* "M-p" 'legit-previous-header)
+(define-key *legit-diff-mode-keymap* "Tab" 'other-window)
+
+;; help
+(define-key lem/peek-legit::*peek-legit-keymap* "?" 'legit-help)
+(define-key lem/peek-legit::*peek-legit-keymap* "C-x ?" 'legit-help)
+;; quit
+(define-key *legit-diff-mode-keymap* "q" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "q" 'legit-quit)
+(define-key *legit-diff-mode-keymap* "M-q" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "M-q" 'peek-legit-quit)
+(define-key *legit-diff-mode-keymap* "Escape" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "Escape" 'peek-legit-quit)
+(define-key *legit-diff-mode-keymap* "C-c C-k" 'legit-quit)
+(define-key lem/peek-legit::*peek-legit-keymap* "C-c C-k" 'legit-quit)
+
+(defun pop-up-message (message)
+  (with-pop-up-typeout-window (s (make-buffer "*legit status*") :erase t)
+    (format s "~a" message)))
+
+(defun last-character (s)
+  (subseq s (- (length s) 2) (- (length s) 1)))
+
+(defmacro with-current-project (&body body)
+  "Execute body with the current working directory changed to the project's root,
+  find and set the VCS system for this operation.
+
+  If no Git directory (or .fossil file) are found, message the user."
+  `(let ((root (lem-core/commands/project:find-root (buffer-directory))))
```

Comment:
It might be a good size to use the call-with auxiliary function, what do you think?